canonical/grafana-k8s-operator
# Use `cos-tool` to inject label matchers

## Issue
Add `cos-tool` to Grafana and the `grafana_dashboard` library. The adaptation of the methods differs slightly from other cases since single, bare expressions will be passed rather than groups inside alerts, and Grafana dashboards
may include variables and negatives which are not generally used in alerts.

In general, this is just passing `--` as a leader to the transformation for `CosTool`

## Solution

In `GrafanaDashboardConsumer`, pass the topology, transformer, and
dashboard content to a new method. This method iterates over panels,
examines expressions, and makes necessary substitutions for
`cos-tool` to parse them, which are then reversed after
injection (re-adding `-` for negative values, re-adding variable
content inside range selectors).


Add reasonably complex unit tests to cover these cases, as well as
bare `cos-tool`



## Testing Instructions
Import a dashboard from the Grafana marketplace into a charm (alternatively, take an unmodified, existing dashboard). Relate it to Grafana and ensure that labels are present.

## Release Notes
Inject topology labels into expressions.

Closes #51
